signature/src/lib.rs

```diff
@@ -30,3 +35,377 @@ pub extern "C" fn verify_rsa(
+    const PROD_HOSTNAME: &str = "aus.content-signature.mozilla.org";
+    // Fetched from: https://aus5.mozilla.org/json/1/FirefoxVPN/2.14.0/WINNT_x86_64/release-cdntest/update.json
+    const PROD_SIGNATURE: &str = "znYFqdKKFgijVgUhnq5VuZxtI5Zay8MARVFr3cG1CbB9eH9slQFkE9ZjMdLzbf5OZqj2gds1OqbCm45L38e2joKD_mCAUGtajebztDdWx9Rqgmn-9vu6t-SCl6HQrzbh";
+    const PROD_INPUT_DATA: &[u8] =
+        b"{\"version\": \"2.17.0\", \"url\": \"https://archive.mozilla.org/pub/vpn/candidates/2.17.0-candidates/build20230926170708/windows/MozillaVPN.msi\", \"required\": false, \"hashFunction\": \"sha512\", \"hashValue\": \"97fe0255e50cd33d3ed50fe94edf3d95b72950b28f9b0157b43195a6c190aaf7e34543044c14d42deabf97b84910da1fbb7f752eaf800130315c2ad5172ea45c\"}";
+
+    #[test]
+    fn test_verify_succeeds_if_valid() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        assert!(r.is_ok(), "Found unexpected error: {}", r.unwrap_err());
+    }
+
+    #[test]
+    fn test_verify_fails_if_root_hash_mismatch() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            &ROOT_HASH.replace("A", "B"),
+            VALID_HOSTNAME,
+        );
+        assert_eq!(r, Err(BalrogError::RootHashMismatch));
+    }
+
+    #[test]
+    fn test_verify_fails_if_cert_invalid() {
+        let r = parse_and_verify(
+            INVALID_CERTIFICATE,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        /* I think this is the wrong error code actually...  */
+        /* I would have expected X509Error::InvalidCertificate */
+        assert_eq!(r, Err(BalrogError::PemDecodeError));
+    }
+
+    #[test]
+    fn test_verify_fails_if_hostname_mismatch() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1615559719, // March 12, 2021
+            ROOT_HASH,
+            "example.com",
+        );
+        assert_eq!(r, Err(BalrogError::HostnameMismatch));
+    }
+
+    #[test]
+    fn test_verify_fails_if_cert_has_expired() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            VALID_SIGNATURE,
+            1215559719, // July 9, 2008
+            ROOT_HASH,
+            VALID_HOSTNAME,
+        );
+        assert_eq!(r, Err(BalrogError::CertificateExpired));
+    }
+
+    #[test]
+    fn test_verify_fails_if_bad_siganture() {
+        let r = parse_and_verify(
+            VALID_CERT_CHAIN,
+            VALID_INPUT,
+            &VALID_SIGNATURE.to_ascii_lowercase(),
```

Comment:
Could use a comment that an invalid signature is being constructed by chaning to lower case.